drivers/power_domain/power_domain_psl.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2022 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_power_domain_psl
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/pm/device.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(power_domain_psl, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+struct npcx_psl_config {
+	struct gpio_dt_spec enable;
+	/* PSL_IN pinmux configuration */
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static int npcx_psl_pm_action(const struct device *dev,
+			     enum pm_device_action action)
+{
+	const struct npcx_psl_config *config = dev->config;
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_OFF:
```

Comment:
I think you should be using suspend action here, check https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/power_domain/power_domain_gpio.c

Why does this PM domain only supports one of the action, can you please explain?

Please check https://docs.zephyrproject.org/latest/services/pm/power_domain.html

Reply:
> I think you should be using suspend action here
> Why does this PM domain only supports one of the action, can you please explain?

Here is  [the table](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nuvoton_npcx/common/power.c#L24) for all supported power states in npcx ec series. Besides ```Stand-By``` power state, soc can enter/leave the other power states  by configuring PMCSR reg in the power management module. 

In order to enter ```Stand-By``` state, we need an external circuit to switch the Core power supply (VCC1) off and gain an ultra-low power consumption. This PR uses ```enable-gpios``` and ```PM_DEVICE_ACTION_TURN_OFF```to trigger it. Then, only a few modules such as PSL still work on the VSBY power rail. It can detect the event from PSL_IN pads (For example, a power button is pressed.), switch on the main power domain of npcx ec, VCC1, power up  the soc, and start to execute the program automatically. That's why I prefer ```turn off``` action since the soc is not powered in ```Stand-By``` state.

Reply:
Add more explanation and hope it is helpful for reviewing. Here is a brief block diagram for using the PSL circuit.
```
	                                  VCC1 Power Rail
	    VCC1 In                              |      
	      +--------------------------+_+-----+
	      |                           |
	+-----+------+-+  PSL_OUT  +------+-------+
	|            | |-----------|   External   |    
	|   NPCX EC  | |           |  PSL Circuit |
	|            | |--PSL_INx  +--------------+
	+------------+++  
	              |       
	        VSBY Power Rail
```
Normally, PSL_OUT is in the ```active``` state that let the external PSL circuit switch on the VCC1 power rail. If npcx ec wants to change its power state to ```TURN_OFF``` (We call it as ``Stand-By`` power state in our datasheet.), it must be triggered by a [power domain driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/pm/device.h#L81) and notify the external PSL circuit (a PM device) by setting PSL_OUT to ```inactive``` state. Then, the PM device will switch off the VCC1 power rail and power down the SoC.

We needn't a ```TURN_ON``` action here because an event from PSL_INx will set PSL_OUT back to ```active``` state automatically. Then, the external PSL circuit will switch on the VCC1 power rail and power the SoC up.

Reply:
Thanks for all the details. I think that you should be using `PM_DEVICE_ACTION_SUSPEND`, though. This domain doesn't seem to require any actions to be turned on/off, like the GPIO one has (enable-gpio seems to be _special_ here). @JordanYates feedback welcome here.

Reply:
@gmarull, thanks for your comments. Whether the action is ```PM_DEVICE_ACTION_SUSPEND``` or ```PM_DEVICE_ACTION_TURN_OFF```, I'm okay with it. But we might need more suggestions here. 

@JordanYates, your feedback is appreciated.

@fabiobaltieri and @keith-zephyr, as users of npcx PSL module in the downstream repo, which action has a better understanding for you?

Reply:
`PM_DEVICE_ACTION_TURN_ON` and `PM_DEVICE_ACTION_TURN_OFF` signifying that the device has lost/gained power.
AFAICT the NPCX EC block is never losing power, so it shouldn't use those actions.
`PM_DEVICE_ACTION_SUSPEND` should trigger `PSL_OUT` to go inactive.
Some other logic will need to tell the driver that it is back in `PM_DEVICE_STATE_ACTIVE` when `PSL_INx`  triggers.

Reply:
>AFAICT the NPCX EC block is never losing power, so it shouldn't use those actions.

@JordanYates . Thanks for your comment. After changing ```PM_DEVICE_ACTION_TURN_OFF``` to ```PM_DEVICE_ACTION_SUSPEND```, I found the behavior is not what I expect. If CONFIG_PM_DEVICE is enabled, the pm will run [PM_DEVICE_ACTION_SUSPEND action](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/pm.c#L78) after CPU left  ```PM_STATE_RUNTIME_IDLE```. It means PSL driver cuts the main power of NPXC EC off when CPU's power state is ```PM_STATE_SUSPEND_TO_IDLE```.

According to the definition of power management states below:
```
PM_STATE_SUSPEND_TO_IDLE:
1. It puts all of the cores in deepest possible idle state
2. No operating state is lost (ie. the cpu core does not lose execution context), so the system can go back to where it left off easily enough.
PM_STATE_SOFT_OFF:
1. It consumes a minimal amount of power
2. The contents of system(CPU and memory) will not be preserved, so the system will be restarted as if from initial power-up and kernel boot.
```

we choose ```Deep Sleep``` and ```Stand-By``` (NPCX EC power state) to implement ```PM_STATE_SUSPEND_TO_IDLE``` and  ```PM_STATE_SOFT_OFF``` power states for power management. Please refer to table 1 and table 2. (When NPCX EC enters '''Stand-By```, CPU and memory will not be preserved anymore.)

Table 1. NPCX EC power state used in Zephyr
``` 
+--------------------------------------------------------------------------+
 | Power State | LFCLK | HFCLK | APB/AHB |  Core  | RAM/Regs   | VCC | VSBY |
 |--------------------------------------------------------------------------|
 | Active      | On    | On    | On      | Active | Active     | On  | On   |
 | Deep Sleep  | On    | Stop  | Stop    | Stop   | Power Down | On  | On   |
 | Stand-By    | Off   | Off   | Off     | Off    | Off        | Off | On   |
 +--------------------------------------------------------------------------+
 ```
 
Table 2. Mapping table between NPCX EC power state/PM State, and Relate device PM action.
```
 +----------------------------------------------------------------------------------+
 | NPCX        | Related     | Supported System         | Device PM Actions         |
 | Power State | Power Module| PM state                 |                           |
 |----------------------------------------------------------------------------------+
 | Active      | Core        | PM_STATE_ACTIVE          | PM_DEVICE_ACTION_RESUME  |
 |----------------------------------------------------------------------------------+
 | Deep Sleep  | PMC         | PM_STATE_SUSPEND_TO_IDLE | PM_DEVICE_ACTION_SUSPEND  |
 |----------------------------------------------------------------------------------+
 | Stand-By    | PSL &       | PM_STATE_SOFT_OFF        | PM_DEVICE_ACTION_TURN_OFF?|
 |             | EXT Circuit |                          |                           |
 |----------------------------------------------------------------------------------+
```

In this PR, PSL and external circuit help NPCX EC enter ```Stand-By``` power state mode. (Most of power used by NPCX is gone in this state. CPU and RAM are not preserved also.) I might misuse ```PM_DEVICE_ACTION_TURN_OFF``` to achieve this purpose.  Is any other action I can use for it? Thanks.

Reply:
Hi @JordanYates and @gmarull:
Any comment about which pm device action is used for PSL power domain driver? If `PM_DEVICE_ACTION_TURN_OFF` still violates the rule of power domain driver, I will close this PR since action `PM_DEVICE_ACTION_SUSPEND` is called every time when CPU enters 'PM_STATE_SUSPEND_TO_IDLE' power state. It causes RAM and CPU are not preserved. That's not what we expect. PSL driver is used only for cases that pursue minimal power consumption.

Reply:
Looks like implementing PSL driver in power domain driver is not a good approach. Closed this PR and create the other [PR](https://github.com/zephyrproject-rtos/zephyr/pull/46428) which only implements how to configure PSL input via pinctrl mechanism.